ParsedFunction can accept another function

andrsd opened the pull request:
Adding the possibility to couple a function into a ParsedFunction, similarly
like we do it for postprocessors and scalar variables

Closes #12179

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

andrsd committed:
ParsedFunction can accept another function

Adding the possibility to couple a function into a ParsedFunction, similarly
like we do it for postprocessors and scalar variables

Closes #12179

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/235390/) on 2e2e17e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/12180/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/12180/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 17aacbd09f60033fb9df0f5f52d75e4802210744`

andrsd commented:
This is good for #MOOSEnewsletter

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/235424/) on 2e11d64 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12180/site)

This comment will be updated on new commits.

permcody commented on framework/src/functions/MooseParsedFunctionWrapper.C:
```diff
@@ -138,16 +139,32 @@ MooseParsedFunctionWrapper::initialize()
     {
       // Use istringstream to convert, if it fails produce an error, otherwise add the variable to
       // the _vals variable
       if (!(ss >> tmp) || !ss.eof())
```
I know you didn't write this code, but we know now that this isn't sufficient to catch all edge cases. We have a utility that should be called here instead. Perhaps you can fix while you are in here?

permcody commented on framework/include/functions/MooseParsedFunctionWrapper.h:
```diff
@@ -104,6 +105,12 @@ class MooseParsedFunctionWrapper
+  std::vector<unsigned int> _fn_index;
+
+  /// Vector of functions
+  std::vector<Function *> _fns;
```
Not my favorite class scoped variable name 😦

andrsd replied:
But, it nicely matches `_pp_vals` and `_pp_index` a few lines above 🙃

andrsd commented:
I actually need to fix a few comments in the header file, because with the fix I made they are now lying.

andrsd committed:
Internal refactoring in MooseParsedFunctionWrapper

1. better variables names, so they are more descriptive
2. fixing typos
3. fixing outdated comments
4. removing cpt. Obvious
5. using MooseUtils::convert for std::string to Real conversion

andrsd commented:
I did some refactoring in the `MooseParsedFunctionWrapper` - see 1bb7c82 for the actual description of things I did. I removed a lot of IMO useless comments from the code, because much better description is now in the header file.
